README.md
[![Unit Tests](https://github.com/xepor/xepor/actions/workflows/test.yml/badge.svg)](https://github.com/xepor/xepor/actions/workflows/test.yml)
[![PyPI-Server](https://img.shields.io/pypi/v/xepor.svg)](https://pypi.org/project/xepor/)
![PyPI - Status](https://img.shields.io/pypi/status/xepor)
[![Documentation Status](https://readthedocs.org/projects/xepor/badge/?version=latest)](https://xepor.readthedocs.io/en/latest/?badge=latest)
[![Project generated with PyScaffold](https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold)](https://pyscaffold.org/)
[![996.icu](https://img.shields.io/badge/link-996.icu-red.svg)](https://996.icu)

# Xepor

[Xepor](https://github.com/xepor/xepor) (pronounced */ˈzɛfə/*, zephyr) is a web routing framework for reverse engineers and security researchers.
It provides a Flask-like API for hackers to intercept and modify HTTP requests and/or HTTP responses in a human-friendly coding style.

This project is meant to be used with [mitmproxy](https://github.com/mitmproxy/mitmproxy/). Users write scripts with `xepor`, and run the script *inside* mitmproxy with `mitmproxy -s your-script.py`.

If you want to step from PoC to production, from demo(e.g. [http-reply-from-proxy.py](https://github.com/mitmproxy/mitmproxy/blob/v7.0.4/examples/addons/http-reply-from-proxy.py), [http-trailers.py](https://github.com/mitmproxy/mitmproxy/blob/v7.0.4/examples/addons/http-trailers.py), [http-stream-modify.py](https://github.com/mitmproxy/mitmproxy/blob/v7.0.4/examples/addons/http-stream-modify.py)) to something you could take out with your WiFi Pineapple, then Xepor is for you!

## Features

1. Code everything with `@api.route()`, just like Flask! Write everything in *one* script and no `if..else` any more.
2. Handle multiple URL routes, even multiple hosts in one `InterceptedAPI` instance.
3. For each route, you can choose to modify the request *before* connecting to server (or even return a fake response without connection to upstream), or modify the response *before* forwarding to user.
4. Blacklist mode or whitelist mode. Only allow URL endpoints defined in scripts to connect to upstream, blocking everything else (in specific domains) with HTTP 404. Suitable for transparent proxying.
5. Human readable URL path definition and matching powered by [parse](https://pypi.org/project/parse/)
6. Host remapping. define rules to redirect to genuine upstream from your fake hosts. Regex matching is supported. **Best for SSL stripping and server-side license cracking**!
7. Plus all the bests from [mitmproxy](https://github.com/mitmproxy/mitmproxy/)! **ALL** operation modes ( `mitmproxy` / `mitmweb` + `regular` / `transparent`  / `socks5` / `reverse:SPEC` / `upstream:SPEC`) are fully supported.

## Use Case

1. Evil AP and phishing through MITM.
2. Sniffing traffic from target device by iptables + transparent proxy, modify the payload with xepor on the fly.
3. Cracking cloud-based software license. See [examples/krisp/](https://github.com/xepor/xepor-examples/tree/main/krisp/) as an example.
4. Write a complicated web crawler in **\~100 lines of code**. See [examples/polyv_scrapper/](https://github.com/xepor/xepor-examples/tree/main/polyv_scrapper/) as an example.
5. ... and many more.

SSL stripping is NOT provided by this project.

# Installation

```bash
pip install xepor
```

# Quick start

Take the script from [examples/httpbin](https://github.com/xepor/xepor-examples/tree/main/httpbin/) as an example.

```bash
mitmweb -s example/httpbin/httpbin.py
```

Set your Browser HTTP Proxy to `http://127.0.0.1:8080`, and access the web interface at http://127.0.0.1:8081/.

Send a GET request from http://httpbin.org/#/HTTP_Methods/get_get , Then you could see the modification made by Xepor in mitmweb interface, browser dev tools or Wireshark.

The `httpbin.py` do two things.

1. When user access http://httpbin.org/get, inject a query string parameter `payload=evil_param` inside HTTP request.
2. When user access http://httpbin.org/basic-auth/xx/xx/ (we just pretend we don't know the password), sniff `Authorization` headers from HTTP requests and print the password to the attacker.

Just what mitmproxy always does, but with code written in [*xepor*](https://github.com/xepor/xepor) way.

```python
# https://github.com/xepor/xepor-examples/tree/main/httpbin/httpbin.py
from mitmproxy.http import HTTPFlow
from xepor import InterceptedAPI, RouteType


HOST_HTTPBIN = "httpbin.org"

api = InterceptedAPI(HOST_HTTPBIN)


@api.route("/get")
def change_your_request(flow: HTTPFlow):
    """
    Modify URL query param.
    Test at:
    http://httpbin.org/#/HTTP_Methods/get_get
    """
    flow.request.query["payload"] = "evil_param"


@api.route("/basic-auth/{usr}/{pwd}", rtype=RouteType.RESPONSE)
def capture_auth(flow: HTTPFlow, usr=None, pwd=None):
    """
    Sniffing password.
    Test at:
    http://httpbin.org/#/Auth/get_basic_auth__user___passwd_
    """
    print(
        f"auth @ {usr} + {pwd}:",
        f"Captured {'successful' if flow.response.status_code < 300 else 'unsuccessful'} login:",
        flow.request.headers.get("Authorization", ""),
    )


addons = [api]
```


CHANGELOG.md
# Changelog

% **Note**: This document contains some [MyST](https://myst-parser.readthedocs.io/en/latest/index.html) syntax which is not completely Markdown. For better experience (and cross references to the changed code) you could read this document on [Read the Docs](https://xepor.readthedocs.io/en/latest/changelog.html)
## Version 0.6.0

- Feature: Automatically set `--set connection_strategy=lazy` when Xepor scripts are loaded. No need to manually set the options any more 🥳
- Feature: Add {func}`xepor.InterceptedAPI.load` API for configuration before start. Check the documentation for usage details.

## Version 0.5.1

- Feature: Add compatibility with mitmproxy 9.x and Python 3.11~3.12.
  - Since mitmproxy 8.x is not compatible with Python 3.11, therefore updating to mitmproxy 9.x + xepor 0.5.1 is recommended especially for Kali users with rolling updates.

## Version 0.5.0

**⚠️Breaking Change⚠️**: Incompatible API changes.

- Change: Move some constants inside {class}`~xepor.InterceptedAPI` to seperate enums: {func}`xepor.RouteType`, {func}`xepor.FlowMeta`. ([xepor/xepor@83128e8](https://github.com/xepor/xepor/commit/83128e81f8c181cd23363d27c76209b9de1c49eb))
- Change: rename `reqtype` in {func}`xepor.InterceptedAPI.route` to `rtype`. The name there before is for historical reason. [xepor/xepor@c6c7e83](https://github.com/xepor/xepor/commit/c6c7e8319aef95de01a0f797d9bb2c01632b7609)
- All exposed API is now type hinted and well documented! check the latest [Documentation](https://xepor.readthedocs.io/en/latest). example scripts are also updated with API change.

## Version 0.4.0

- Feature: Add compatibility with mitmproxy 8.x.

## Version 0.3.0

- Feature: Add `return_error` option for {func}`xepor.InterceptedAPI.route` ([xepor/xepor@fddad13](https://github.com/xepor/xepor/commit/fddad13efc0ed08b214c4f4078e130b04deeeab0))
- Add more docs and examples. Also some tests to ensure the script stay up-to-date with version changes. ([xepor/xepor@3f96bf1](https://github.com/xepor/xepor/commit/3f96bf16b8d45b9fb49485abd475b7139422e1b9))

## Version 0.2.0

**⚠️Breaking Change⚠️**: mitmproxy 7.0 contains backward-incompatible API change comparing with <=6.x, update to 7.x may need to modify all of your scripts (including those not written with Xepor). Just a change on some methods' signatures.

- Feature: Add `blacklist_domain` and `respect_proxy_headers` options for {class}`~xepor.InterceptedAPI`. Refer to [API Doc](https://xepor.readthedocs.io/en/latest/api/modules.html#xepor.InterceptedAPI) for their usage. ([xepor/xepor@aa2517d](https://github.com/xepor/xepor/commit/aa2517d65f800944587a25678b1e624f84bdf13f))
- Feature: Add a debug header when client request hitting {func}`~xepor.InterceptedAPI.default_response` ([xepor/xepor@d57be79](https://github.com/xepor/xepor/commit/d57be794a2e8863bc8aebec070f4a78be8ec121b))
- Fix: adopt API change in mitmproxy 7.x ([xepor/xepor@beedeab](https://github.com/xepor/xepor/commit/beedeab446508bf0df82c90e05d99b5da9ee3753))
- Fix: Request is forwarded to wrong destination host when mitmproxy is start in reverse/upstream mode. ([xepor/xepor@aa2517d](https://github.com/xepor/xepor/commit/aa2517d65f800944587a25678b1e624f84bdf13f)) **Be sure to start mitmproxy with `--set connection_strategy=lazy` under ANY circumstances or the upstream connection may still be wrong.**
- Update example scripts.

## Version 0.1 (development)

This is **the only version** compatible with mitmproxy <= 6.x

- Implement most features.


tests/tserver.py
"""
Imported from:
https://github1s.com/mitmproxy/mitmproxy/blob/v7.0.4/test/mitmproxy/tservers.py
"""
from unittest import mock

from mitmproxy import controller
from mitmproxy import eventsequence
from mitmproxy import io
from mitmproxy.proxy import server_hooks
from mitmproxy.test import tflow
from mitmproxy.test import tutils


class MasterTest:

    async def cycle(self, master, content):
        f = tflow.tflow(req=tutils.treq(content=content))
        layer = mock.Mock("mitmproxy.proxy.protocol.base.Layer")
        layer.client_conn = f.client_conn
        layer.reply = controller.DummyReply()
        await master.addons.handle_lifecycle(server_hooks.ClientConnectedHook(layer))
        for e in eventsequence.iterate(f):
            await master.addons.handle_lifecycle(e)
        await master.addons.handle_lifecycle(server_hooks.ClientDisconnectedHook(layer))
        return f

    async def dummy_cycle(self, master, n, content):
        for i in range(n):
            await self.cycle(master, content)
        await master._shutdown()

    def flowfile(self, path):
        with open(path, "wb") as f:
            fw = io.FlowWriter(f)
            t = tflow.tflow(resp=True)
            fw.add(t)


tests/conftest.py
"""
    Dummy conftest.py for xepor.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html

    Imported from:
    https://github1s.com/mitmproxy/mitmproxy/blob/v7.0.4/test/conftest.py
"""

import os
import socket

from mitmproxy.options import Options
from mitmproxy.utils import data

import pytest

# pytest_plugins = ('test.full_coverage_plugin',)

skip_windows = pytest.mark.skipif(
    os.name == "nt",
    reason='Skipping due to Windows'
)

skip_not_windows = pytest.mark.skipif(
    os.name != "nt",
    reason='Skipping due to not Windows'
)

try:
    s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
    s.bind(("::1", 0))
    s.close()
except OSError:
    no_ipv6 = True
else:
    no_ipv6 = False

skip_no_ipv6 = pytest.mark.skipif(
    no_ipv6,
    reason='Host has no IPv6 support'
)


@pytest.fixture()
def tdata():
    return data.Data(__name__)

@pytest.fixture()
def toptions():
    options = Options()
    # options duplicated here to simplify testing setup
    options.add_option(
        "connection_strategy",
        str,
        "lazy",
        "Determine when server connections should be established.",
        choices=("eager", "lazy"),
    )
    options.add_option(
        "keep_host_header",
        bool,
        False,
        """
        Reverse Proxy: Keep the original host header instead of rewriting it
        to the reverse proxy target.
        """,
    )
    return options


tests/test_single.py
import pytest
from mitmproxy.http import HTTPFlow, Response
from mitmproxy.test import taddons, tflow
from xepor import InterceptedAPI


__author__ = "ttimasdf"
__copyright__ = "ttimasdf"
__license__ = "Apache-2.0"


@pytest.fixture
def api_simple():
    api = InterceptedAPI("example.com")

    @api.route("/test")
    def route1(flow: HTTPFlow):
        flow.response = Response.make(200, "TEST intercepted")

    @api.route(
        "/{}/{}/{vid}_1.m3u8",
        "hls.videocc.net",
    )
    def route2(flow: HTTPFlow, *args, vid):
        flow.response = Response.make(200, "TEST 2 INTERCEPTED")

    return api


@pytest.mark.parametrize(
    "req_url,resp_body",
    [
        ("http://example.com/test", "TEST intercepted"),
        ("http://hls.videocc.net/jkag324wd2/e/cwqzcxkvj0iukomqxu0l591u2dke4vkc_1.m3u8", "TEST 2 INTERCEPTED"),
    ],
)
def test_intercepted_route(toptions, api_simple, req_url, resp_body):
    with taddons.context(api_simple, options=toptions) as tctx:
        flow = tflow.tflow()
        flow.request.url = req_url
        assert flow.response is None

        api_simple.request(flow)
        assert resp_body in flow.response.text


def test_non_intercepted_route(toptions, api_simple):
    with taddons.context(api_simple, options=toptions) as tctx:
        flow = tflow.tflow()
        flow.request.path = "/test2"
        assert flow.response is None

        api_simple.request(flow)
        assert flow.response is None


src/xepor/skeleton.py
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         fibonacci = xepor.skeleton:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This skeleton file can be safely removed if not needed!

References:
    - https://setuptools.pypa.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys

from xepor import __version__

__author__ = "ttimasdf"
__copyright__ = "ttimasdf"
__license__ = "Apache-2.0"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from xepor.skeleton import fib`,
# when using this Python module as a library.


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for _i in range(n - 1):
        a, b = b, a + b
    return a


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="xepor {ver}".format(ver=__version__),
    )
    parser.add_argument(dest="n", help="n-th Fibonacci number", type=int, metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(args):
    """Wrapper allowing :func:`fib` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`fib`, it prints the result to the
    ``stdout`` in a nicely formatted message.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m xepor.skeleton 42
    #
    run()


src/xepor/xepor.py
import functools
import logging
import os
import re
import sys
import traceback
import urllib.parse
from enum import Enum
from typing import List, Optional, Tuple, Union

from mitmproxy import ctx
from mitmproxy.addonmanager import Loader
from mitmproxy.connection import Server
from mitmproxy.http import HTTPFlow, Response
from mitmproxy.net.http import url
from parse import Parser

__author__ = "ttimasdf"
__copyright__ = "ttimasdf"
__license__ = "Apache-2.0"


class RouteType(Enum):
    """
    This enum is an option set in route definition, specify it to be matched
    on either incoming request or response.
    """

    REQUEST = 1
    """The route will be matched on mitmproxy ``request`` event"""
    RESPONSE = 2
    """The route will be matched on mitmproxy ``response`` event"""


class FlowMeta(Enum):
    """
    This class is used internally by Xepor to mark ``flow`` object by certain metadata.
    Refer to the source code for detailed usage.
    """

    REQ_PASSTHROUGH = "xepor-request-passthrough"
    RESP_PASSTHROUGH = "xepor-response-passthrough"
    REQ_URLPARSE = "xepor-request-urlparse"
    REQ_HOST = "xepor-request-host"


class InterceptedAPI:
    """
    the InterceptedAPI object is the central registry of your view functions.
    Users should use a function decorator :func:`route` to define and register
    URL and host mapping to the view functions. Just like flask's :external:py:meth:`flask.Flask.route`.

    .. code-block:: python

        from xepor import InterceptedAPI, RouteType

        HOST_HTTPBIN = "httpbin.org"
        api = InterceptedAPI(HOST_HTTPBIN)

    Defining a constant for your target (victim) domain name is not mandatory
    (even the `default_host` parameter itself is optional) but
    recommanded as a best practise. If you have multiple hosts to inject
    (see an example at `xepor/xepor-examples/polyv_scrapper/polyv.py <https://github.com/xepor/xepor-examples/blob/306ffad36a9ff3db00eb44b67b8b83a85e234d6e/polyv_scrapper/polyv.py#L27-L29>`_), you would have to specify the domain name
    multiple times in each :func:`route` in `host` parameter,
    (especially for domains other than `default_host`).
    So it's better to have a variable for that.

    Add route via function call similar to Flask :external:py:meth:`flask.Flask.add_url_rule`
    is not yet implemented.

    :param default_host: The default host to forward requests to.
    :param host_mapping: A list of tuples of the form (regex, host) where regex
        is a regular expression to match against the request host and host is the
        host to redirect the request to.
    :param blacklist_domain: A list of domains to not forward requests to.
        The requests and response from hosts in this list will not respect
        :py:obj:`request_passthrough` and :py:obj:`response_passthrough` setting.
    :param request_passthrough: Whether to forward the request to upstream server
        if no route is found. If ``request_passthrough = False``, all requests not
        matching any route will be responded with :func:`default_response` without
        connecting to upstream.
    :param response_passthrough: Whether to forward the response to the user
        if no route is found. If ``response_passthrough = False``, all responses not
        matching any route will be replaced with the Response object
        generated by :func:`default_response`.
    :param respect_proxy_headers: Set to `True` only when you use Xepor as
        a web server behind a reverse proxy. Typical use case is to set up an
        mitmproxy in ``reverse`` mode to bypass some online license checks.
        Xepor will respect the following headers and strip them from requests to upstream.

        - `X-Forwarded-For`
        - `X-Forwarded-Host`
        - `X-Forwarded-Port`
        - `X-Forwarded-Proto`
        - `X-Forwarded-Server`
        - `X-Real-Ip`
    """

    _REGEX_HOST_HEADER = re.compile(r"^(?P<host>[^:]+|\[.+\])(?::(?P<port>\d+))?$")

    _PROXY_FORWARDED_HEADERS = [
        "X-Forwarded-For",
        "X-Forwarded-Host",
        "X-Forwarded-Port",
        "X-Forwarded-Proto",
        "X-Forwarded-Server",
        "X-Real-Ip",
    ]

    def __init__(
        self,
        default_host: Optional[str] = None,
        host_mapping: List[Tuple[Union[str, re.Pattern], str]] = {},
        blacklist_domain: List[str] = [],
        request_passthrough: bool = True,
        response_passthrough: bool = True,
        respect_proxy_headers: bool = False,
    ):

        self.default_host = default_host
        self.host_mapping = host_mapping
        self.request_routes: List[Tuple[Optional[str], Parser, callable]] = []
        self.response_routes: List[Tuple[Optional[str], Parser, callable]] = []
        self.blacklist_domain = blacklist_domain
        self.request_passthrough = request_passthrough
        self.response_passthrough = response_passthrough
        self.respect_proxy_headers = respect_proxy_headers

        self._log = logging.getLogger(__name__)
        if os.getenv("XEPOR_LOG_DEBUG"):
            self._log.setLevel(logging.DEBUG)
        self._log.info("%s started", self.__class__.__name__)

    def load(self, loader: Loader):
        """
        This function is called by the mitmproxy framework *before* proxy server started.
        Currently it's used to set a must-have mitmproxy option for Xepor
        to work: ``connection_strategy=lazy``. If you want to override this method,
        remember to call ``super().load(loader)`` in your code.

        User can also import and use :py:data:`mitmproxy.ctx` object to configure
        other options for mitmproxy when overriding this function.

        .. code-block:: python

            from mitmproxy import ctx

            ctx.options.connection_strategy = "lazy"

        :param loader: a :py:class:`mitmproxy.addonmanager.Loader` which can be used to add custom options.
        :return: None
        """
        self._log.info("Setting option connection_strategy=lazy")
        ctx.options.connection_strategy = "lazy"

    # def server_connect(self, data: ServerConnectionHookData):
    #     self._log.debug("Getting connection: peer=%s sock=%s addr=%s . state=%s",
    #         data.server.peername, data.server.sockname, data.server.address, data.server)

    def request(self, flow: HTTPFlow):
        """
        This function is called by the mitmproxy framework whenever a request is made.

        :param flow: The :py:class:`mitmproxy.http.HTTPFlow` object from client request.
        :return: None
        """
        if FlowMeta.REQ_URLPARSE in flow.metadata:
            url = flow.metadata[FlowMeta.REQ_URLPARSE]
        else:
            url = urllib.parse.urlparse(flow.request.path)
            flow.metadata[FlowMeta.REQ_URLPARSE] = url
        path = url.path
        if flow.metadata.get(FlowMeta.REQ_PASSTHROUGH) is True:
            self._log.warning(
                "<= [%s] %s skipped because of previous passthrough",
                flow.request.method,
                path,
            )
            return
        host = self.remap_host(flow)
        handler, params = self.find_handler(host, path, RouteType.REQUEST)
        if handler is not None:
            self._log.info("<= [%s] %s", flow.request.method, path)
            handler(flow, *params.fixed, **params.named)
        elif (
            not self.request_passthrough
            or self.get_host(flow)[0] in self.blacklist_domain
        ):
            self._log.warning("<= [%s] %s default response", flow.request.method, path)
            flow.response = self.default_response()
        else:
            flow.metadata[FlowMeta.REQ_PASSTHROUGH] = True
            self._log.debug("<= [%s] %s passthrough", flow.request.method, path)

    def response(self, flow: HTTPFlow):
        """
        This function is called by the mitmproxy when a response is returned the server.

        :param flow: The :py:class:`mitmproxy.http.HTTPFlow` object from server response.
        :return: None
        """
        if FlowMeta.REQ_URLPARSE in flow.metadata:
            url = flow.metadata[FlowMeta.REQ_URLPARSE]
        else:
            url = urllib.parse.urlparse(flow.request.path)
            flow.metadata[FlowMeta.REQ_URLPARSE] = url
        path = url.path
        if flow.metadata.get(FlowMeta.RESP_PASSTHROUGH) is True:
            self._log.warning(
                "=> [%s] %s skipped because of previous passthrough",
                flow.response.status_code,
                path,
            )
            return
        handler, params = self.find_handler(
            self.get_host(flow)[0], path, RouteType.RESPONSE
        )
        if handler is not None:
            self._log.info("=> [%s] %s", flow.response.status_code, path)
            handler(flow, *params.fixed, **params.named)
        elif (
            not self.response_passthrough
            or self.get_host(flow)[0] in self.blacklist_domain
        ):
            self._log.warning(
                "=> [%s] %s default response", flow.response.status_code, path
            )
            flow.response = self.default_response()
        else:
            flow.metadata[FlowMeta.RESP_PASSTHROUGH] = True
            self._log.debug("=> [%s] %s passthrough", flow.response.status_code, path)

    def route(
        self: str,
        path: str,
        host: Optional[str] = None,
        rtype: RouteType = RouteType.REQUEST,
        catch_error: bool = True,
        return_error: bool = False,
    ):
        """
        This is the main API used by end users.
        It decorate a view function to register it with given host and URL.

        Typical usage (taken from official example: `httpbin.py <https://github.com/xepor/xepor-examples/blob/main/httpbin/httpbin.py>`_):

        .. code-block:: python

            @api.route("/get")
            def change_your_request(flow: HTTPFlow):
                flow.request.query["payload"] = "evil_param"

            @api.route("/basic-auth/{usr}/{pwd}", rtype=RouteType.RESPONSE)
            def capture_auth(flow: HTTPFlow, usr=None, pwd=None):
                print(
                    f"auth @ {usr} + {pwd}:",
                    f"Captured {'successful' if flow.response.status_code < 300 else 'unsuccessful'} login:",
                    flow.request.headers.get("Authorization", ""),
                )

        See Github: `xepor/xepor-examples <https://github.com/xepor/xepor-examples>`_ for more examples.


        :param path: The URL path to be routed.
            The path definition grammar is similar to Python 3 :py:func:`str.format`.
            Check the documentation of ``parse`` library:
            `r1chardj0n3s/parse <https://github.com/r1chardj0n3s/parse>`_

        :param host: The host to be routed.
            This value will be matched against the following fields of
            incoming flow object by order:

            1. ``X-Forwarded-For`` Header. (only when `respect_proxy_headers` in :class:`InterceptedAPI` is `True`)
            2. HTTP ``Host`` Header, if exists.
            3.  ``flow.host`` reported by underlying layer.
                In HTTP or Socks5h proxy mode, it may hopefully be a hostname,
                otherwise, it'll be an IP address.

        :param rtype: Set the route be matched on either request or response.
            Accepting :class:`RouteType`.

        :param catch_error: If set to `True`, the exception inside the route
            will be handled by Xepor.

            If set to `False`, the exception will be raised and handled by mitmproxy.

        :param return_error: If set to `True`, the error message inside the exception
            (``str(exc)``) will be returned to client. This behaviour can be overrided
            through :func:`error_response`.

            If set to `False`, the exception will be printed to console,
            the ``flow`` object will be passed to mitmproxy continuely.

            .. admonition:: Note

                When exception occured, the ``flow`` object do `not` always stay intact.
                This option is only a try-catch like normal Python code. If you run
                ``modify1(flow) and modify2(flow) and modify3(flow)`` and exception raised
                in ``modify2()``, the ``flow`` object will be modified partially.

        :return: The decorated function.
        """
        host = host or self.default_host

        def catcher(func):
            """
            The internal wrapper for catching exceptions
            if `catch_error` is specified.
            """

            @functools.wraps(func)
            def handler(flow: HTTPFlow, *args, **kwargs):
                try:
                    return func(flow, *args, **kwargs)
                except Exception as e:
                    etype, value, tback = sys.exc_info()
                    tb = "".join(traceback.format_exception(etype, value, tback))
                    self._log.error(
                        "Exception catched when handling response to %s:\n%s",
                        flow.request.pretty_url,
                        tb,
                    )
                    if return_error:
                        flow.response = self.error_response(str(e))

            return handler

        def wrapper(handler):
            if catch_error:
                handler = catcher(handler)
            if rtype == RouteType.REQUEST:
                self.request_routes.append((host, Parser(path), handler))
            elif rtype == RouteType.RESPONSE:
                self.response_routes.append((host, Parser(path), handler))
            else:
                raise ValueError(f"Invalid route type: {rtype}")
            return handler

        return wrapper

    def remap_host(self, flow: HTTPFlow, overwrite=True):
        """
        Remaps the host of the flow to the destination host.

        .. admonition:: Note

            This function is used internally by Xepor.
            Refer to the source code for customization.

        :param flow: The flow to remap.
        :param overwrite: Whether to overwrite the host and port of the flow.
        :return: The remapped host.
        """
        host, port = self.get_host(flow)
        for src, dest in self.host_mapping:
            if (isinstance(src, re.Pattern) and src.match(host)) or (
                isinstance(src, str) and host == src
            ):
                if overwrite and (
                    flow.request.host != dest or flow.request.port != port
                ):
                    if self.respect_proxy_headers:
                        flow.request.scheme = flow.request.headers["X-Forwarded-Proto"]
                    flow.server_conn = Server((dest, port))
                    flow.request.host = dest
                    flow.request.port = port
                self._log.debug(
                    "flow: %s, remapping host: %s -> %s, port: %d",
                    flow,
                    host,
                    dest,
                    port,
                )
                return dest
        return host

    def get_host(self, flow: HTTPFlow) -> Tuple[str, int]:
        """
        Gets the host and port of the request.
        Extending from :py:attr:`mitmproxy.http.HTTPFlow.pretty_host` to accept
        values from proxy headers(``X-Forwarded-Host`` and ``X-Forwarded-Port``)

        .. admonition:: Note

            This function is used internally by Xepor.
            Refer to the source code for customization.

        :param flow: The HTTPFlow object.
        :return: A tuple of the host and port.
        """
        if FlowMeta.REQ_HOST not in flow.metadata:
            if self.respect_proxy_headers:
                # all(h in flow.request.headers for h in ["X-Forwarded-Host", "X-Forwarded-Port"])
                host = flow.request.headers["X-Forwarded-Host"]
                port = int(flow.request.headers["X-Forwarded-Port"])
            else:
                # Get Destination Host
                host, port = url.parse_authority(flow.request.pretty_host, check=False)
                port = port or url.default_port(flow.request.scheme) or 80
            flow.metadata[FlowMeta.REQ_HOST] = (host, port)
        return flow.metadata[FlowMeta.REQ_HOST]

    def default_response(self):
        """
        This is the default response function for Xepor.
        It will be called in following conditions:

        1. target host in HTTP request matches the ones in `blacklist_domain`.
        2. either `request_passthrough` or `response_passthrough` is set to `False`,
           and no route matches the incoming flow.

        Override this function if it suits your needs.

        :return: A Response object with status code 404
            and HTTP header ``X-Intercepted-By`` set to ``xepor``.
        """
        return Response.make(404, "Not Found", {"X-Intercepted-By": "xepor"})

    def error_response(self, msg: str = "APIServer Error"):
        """
        Returns a response with status code 502 and custom error message.

        Override this function if it suits your needs.

        :param msg: The message to be returned.

        :return: A Response object with status code 502
            and content set to the .
        """
        return Response.make(502, msg)

    def find_handler(self, host, path, rtype=RouteType.REQUEST):
        """
        Finds the appropriate handler for the request.

        .. admonition:: Note

            This function is used internally by Xepor.
            Refer to the source code for customization.

        :param host: The host of the request.
        :param path: The path of the request.
        :param rtype: The type of the route. Accepting :class:`RouteType`.
        :return: The handler and the parse result.
        """
        if rtype == RouteType.REQUEST:
            routes = self.request_routes
        elif rtype == RouteType.RESPONSE:
            routes = self.response_routes
        else:
            raise ValueError(f"Invalid route type: {rtype}")

        for h, parser, handler in routes:
            if h != host:
                continue
            parse_result = parser.parse(path)
            self._log.debug("Parse %s => %s", path, parse_result)
            if parse_result is not None:
                return handler, parse_result

        return None, None
